circuits/update_state_tree.circom

```diff
@@ -1,106 +1,266 @@
+  component vote_options_hash = Hasher(1);
+  vote_options_hash.key <== 0;
+  vote_options_hash.in[0] <== vote_options_leaf_raw;
+  
+  component vote_options_tree_valid = LeafExists(vote_options_tree_depth);
+  vote_options_tree_valid.root <== vote_options_tree_root;
+  vote_options_tree_valid.leaf <== vote_options_hash.hash;
+  for (var i = 0; i < vote_options_tree_depth; i++) {
+    vote_options_tree_valid.path_elements[i] <== vote_options_tree_path_elements[i];
+    vote_options_tree_valid.path_index[i] <== vote_options_tree_path_index[i];
+  }
+
+  // Update vote_option_tree_root with newly updated vote weight
+  component new_vote_options_leaf = Hasher(1);
+  new_vote_options_leaf.key <== 0;
+  new_vote_options_leaf.in[0] <== vote_options_leaf_raw + decrypted_command.out[4]
+
+  component new_vote_options_tree = MerkleTreeUpdate(vote_options_tree_depth);
+  new_vote_options_tree.leaf <== new_vote_options_leaf.hash;
+  for (var i = 0; i < vote_options_tree_depth; i++) {
+    new_vote_options_tree.path_elements[i] <== vote_options_tree_path_elements[i];
+    new_vote_options_tree.path_index[i] <== vote_options_tree_path_index[i];
+  }
+
+  // Verify signature against existing public key
+  component signature_verifier = VerifySignature(message_length - 4);
+
+  signature_verifier.from_x <== state_tree_data[0]; // public key x
+  signature_verifier.from_y <== state_tree_data[1]; // public key y
+
+  signature_verifier.R8x <== decrypted_command.out[message_length - 4]; // sig R8x
+  signature_verifier.R8y <== decrypted_command.out[message_length - 3]; // sig R8x
+  signature_verifier.S <== decrypted_command.out[message_length - 2]; // sig S
+
+  for (var i = 0; i < message_length - 4; i++) {
+    signature_verifier.preimage[i] <== decrypted_command.out[i];
+  }
+
+  // Calculate new vote credits
+  var new_vote_credits = state_tree_data[3] - (decrypted_command.out[4] * decrypted_command.out[4]);
+
+  // Construct new state tree data (and its hash)
+  var new_state_tree_data = [
+    decrypted_command.out[1],
+    decrypted_command.out[2],
+    new_vote_options_tree.root,
+    new_vote_credits,
+    decrypted_command.out[5]
+  ];
+
+  component new_state_tree_leaf = Hasher(state_tree_data_length);
+  new_state_tree_leaf.key <== 0;
+  for (var i = 0; i < state_tree_data_length; i++) {
+    new_state_tree_leaf.in[i] <== new_state_tree_data[i];
   }
 
-  // **** 3.2 Validate signature against existing public key **** //
-  component signature_verifier = VerifySignature(message_length);
+  // Checks to see if we should update the
+  // random leaf (Note: 0 is true, anything non-zero is false)
+  signal valid_signature;
+  valid_signature <-- signature_verifier.valid == 0 ? 0 : 1;
 
-  signature_verifier.from_x <== existing_public_key[0]; // public key x
-  signature_verifier.from_y <== existing_public_key[1]; // public key y
+  signal sufficient_vote_credits;
+  sufficient_vote_credits <-- new_vote_credits > 0 ? 0 : 1;
```

Comment:
This doesn't generate a constraint, `IsZero` should be used here. In this case, you also want `> 0`, which probably means you want to limit the range of `new_vote_credits` (e.g., 32 bits).